### PR TITLE
fix(file): Always create directories after processing path variables

### LINKED ISSUE #10387: bug: file destination doesn't create folders properly
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

With a spec including this:

```
    path: "path/to/files/{{TABLE}}/{{UUID}}.{{FORMAT}}"
```

The `file` destination plugin creates a folder literally named `path/to/files/{{TABLE}}` (not substituting the table name), but does not put any files in it.

The log includes a line like:

```
2023-04-27T15:47:50Z ERR failed to write batch error="open path/to/files/hackernews_items/5b6a1dfb-2375-43b9-a669-e7f8163a7327.csv: no such file or directory" duration=0.01859 len=31 module=cli table=hackernews_items
```

I suspect the variable substitution is not happening at the point where the folder is created, and it should be (so it creates the folder with the literal `{{TABLE}}` token in it, and then tries to write to a location with the table name substituted, but then fails when the folder doesn't exist.

If I manually create the full path (eg `path/to/files/hackernews_items`) then the data is placed there as expected.

### Expected Behavior

The plugin should create the full path(s) it needs to create.

### CloudQuery (redacted) config

```yaml
kind: source
spec:
  name: "hackernews"
  path: "cloudquery/hackernews"
  version: "v1.3.0"
  tables: ["*"]
  backend: local
  destinations:
    - "file"
  spec:
    item_concurrency: 100
    start_time: "2023-04-26T00:00:00Z"
---
kind: destination
spec:
  name: "file"
  path: "cloudquery/file"
  version: "v3.0.0"
  write_mode: "append" # file only supports 'append' mode
  spec:
    path: "path/to/files/{{TABLE}}/{{UUID}}.{{FORMAT}}"
    format: "csv" # supported values are 'csv', 'json' and 'parquet'
```

### Steps To Reproduce

1.  Run `cloudquery sync ./test_config.yml`
2. Observe `ls -la path/to/files/ `

### CloudQuery (redacted) logs

```
2023-04-27T16:20:27Z INF Loading spec(s) args=["./test_config.yml"] module=cli
2023-04-27T16:20:28Z INF Source plugin server listening address=/tmp/cq-doFdwiaueiZONRZD.sock module=cli
2023-04-27T16:20:29Z INF Start sync destinations=["file (v3.0.0)"] module=cli source="hackernews (v1.3.0)" sync_time=2023-04-27T16:20:29Z
2023-04-27T16:20:29Z INF Destination plugin server listening address=/tmp/cq-pKMAjryIHiSHDgtE.sock module=cli
2023-04-27T16:20:30Z INF started call grpc.code=OK grpc.component=server grpc.method=Configure grpc.method_type=unary grpc.service=proto.Destination grpc.start_time=2023-04-27T12:20:30-04:00 grpc.time_ms=0.039 module=cli peer.address=@ protocol=grpc
2023-04-27T16:20:30Z INF finished call grpc.code=OK grpc.component=server grpc.method=Configure grpc.method_type=unary grpc.service=proto.Destination grpc.start_time=2023-04-27T12:20:30-04:00 grpc.time_ms=0.365 module=cli peer.address=@ protocol=grpc
2023-04-27T16:20:30Z INF started call grpc.code=OK grpc.component=server grpc.method=Init grpc.method_type=unary grpc.service=cloudquery.source.v1.Source grpc.start_time=2023-04-27T12:20:30-04:00 grpc.time_ms=0.029 module=cli peer.address=@ protocol=grpc
2023-04-27T16:20:30Z INF finished call grpc.code=OK grpc.component=server grpc.method=Init grpc.method_type=unary grpc.service=cloudquery.source.v1.Source grpc.start_time=2023-04-27T12:20:30-04:00 grpc.time_ms=0.299 module=cli peer.address=@ protocol=grpc
2023-04-27T16:20:30Z INF started call grpc.code=OK grpc.component=server grpc.method=GetDynamicTables grpc.method_type=unary grpc.service=cloudquery.source.v1.Source grpc.start_time=2023-04-27T12:20:30-04:00 grpc.time_ms=0.022 module=cli peer.address=@ protocol=grpc
2023-04-27T16:20:30Z INF finished call grpc.code=OK grpc.component=server grpc.method=GetDynamicTables grpc.method_type=unary grpc.service=cloudquery.source.v1.Source grpc.start_time=2023-04-27T12:20:30-04:00 grpc.time_ms=0.341 module=cli peer.address=@ protocol=grpc
2023-04-27T16:20:30Z INF Start migration destinations=["file (v3.0.0)"] module=cli source="hackernews (v1.3.0)"
2023-04-27T16:20:30Z INF started call grpc.code=OK grpc.component=server grpc.method=Migrate grpc.method_type=unary grpc.service=proto.Destination grpc.start_time=2023-04-27T12:20:30-04:00 grpc.time_ms=0.009 module=cli peer.address=@ protocol=grpc
2023-04-27T16:20:30Z INF finished call grpc.code=OK grpc.component=server grpc.method=Migrate grpc.method_type=unary grpc.service=proto.Destination grpc.start_time=2023-04-27T12:20:30-04:00 grpc.time_ms=0.132 module=cli peer.address=@ protocol=grpc
2023-04-27T16:20:30Z INF End migration destinations=["file (v3.0.0)"] module=cli num_tables=1 source="hackernews (v1.3.0)" time_took=0.000479147
2023-04-27T16:20:30Z INF Start fetching resources destinations=["file (v3.0.0)"] module=cli source="hackernews (v1.3.0)"
2023-04-27T16:20:30Z INF started call grpc.code=OK grpc.component=server grpc.method=Sync grpc.method_type=server_stream grpc.service=cloudquery.source.v1.Source grpc.start_time=2023-04-27T12:20:30-04:00 grpc.time_ms=0.016 module=cli peer.address=@ protocol=grpc
2023-04-27T16:20:30Z INF started call grpc.code=OK grpc.component=server grpc.method=Write2 grpc.method_type=client_stream grpc.service=proto.Destination grpc.start_time=2023-04-27T12:20:30-04:00 grpc.time_ms=0.066 module=cli peer.address=@ protocol=grpc
2023-04-27T16:20:30Z INF top level table resolver started client=hackernews module=hackernews-src table=hackernews_items
2023-04-27T16:20:30Z INF No previous cursor found module=hackernews-src
2023-04-27T16:20:30Z INF Found max ID max_id=35730409 module=hackernews-src
2023-04-27T16:20:30Z INF Finding first post after start_time module=hackernews-src start_time=2023-04-26T00:00:00Z
2023-04-27T16:20:31Z INF Found first post after start_time module=hackernews-src start_item_id=35708284 start_time=2023-04-26T00:00:00Z
2023-04-27T16:20:31Z INF Start item ID is after cursor, updating cursor module=hackernews-src new_cursor=35708284 old_cursor=0
2023-04-27T16:20:40Z ERR failed to write batch error="open path/to/files/hackernews_items/668adbff-03c1-4968-b76c-ca58f04f8255.csv: no such file or directory" duration=0.043906 len=9396 module=cli table=hackernews_items
2023-04-27T16:20:47Z ERR failed to write batch error="open path/to/files/hackernews_items/abd9639f-45c9-4f88-aa2b-4da147b4507b.csv: no such file or directory" duration=0.021857 len=9125 module=cli table=hackernews_items
2023-04-27T16:20:51Z INF table sync finished client=hackernews errors=0 module=hackernews-src resources=22125 table=hackernews_items
2023-04-27T16:20:51Z INF sync finished duration=20794.17992 errors=0 module=hackernews-src panics=0 resources=22125
2023-04-27T16:20:51Z INF finished call grpc.code=OK grpc.component=server grpc.method=Sync grpc.method_type=server_stream grpc.service=cloudquery.source.v1.Source grpc.start_time=2023-04-27T12:20:30-04:00 grpc.time_ms=20794.297 module=cli peer.address=@ protocol=grpc
2023-04-27T16:20:51Z ERR failed to write batch error="open path/to/files/hackernews_items/f295b516-a6a4-40f6-96fb-234308004625.csv: no such file or directory" duration=0.021172 len=3604 module=cli table=hackernews_items
2023-04-27T16:20:51Z INF finished call grpc.code=OK grpc.component=server grpc.method=Write2 grpc.method_type=client_stream grpc.service=proto.Destination grpc.start_time=2023-04-27T12:20:30-04:00 grpc.time_ms=20803.305 module=cli peer.address=@ protocol=grpc
2023-04-27T16:20:51Z INF started call grpc.code=OK grpc.component=server grpc.method=Close grpc.method_type=unary grpc.service=proto.Destination grpc.start_time=2023-04-27T12:20:51-04:00 grpc.time_ms=0.012 module=cli peer.address=@ protocol=grpc
2023-04-27T16:20:51Z INF finished call grpc.code=OK grpc.component=server grpc.method=Close grpc.method_type=unary grpc.service=proto.Destination grpc.start_time=2023-04-27T12:20:51-04:00 grpc.time_ms=0.017 module=cli peer.address=@ protocol=grpc
2023-04-27T16:20:51Z INF started call grpc.code=OK grpc.component=server grpc.method=GetMetrics grpc.method_type=unary grpc.service=cloudquery.source.v1.Source grpc.start_time=2023-04-27T12:20:51-04:00 grpc.time_ms=0.011 module=cli peer.address=@ protocol=grpc
2023-04-27T16:20:51Z INF finished call grpc.code=OK grpc.component=server grpc.method=GetMetrics grpc.method_type=unary grpc.service=cloudquery.source.v1.Source grpc.start_time=2023-04-27T12:20:51-04:00 grpc.time_ms=0.067 module=cli peer.address=@ protocol=grpc
2023-04-27T16:20:51Z INF waiting for destination plugin to terminate module=cli
2023-04-27T16:20:51Z INF Got stop signal. Destination plugin server shutting down address=/tmp/cq-pKMAjryIHiSHDgtE.sock module=cli signal=interrupt
2023-04-27T16:20:51Z INF End sync destinations=["file (v3.0.0)"] module=cli source="hackernews (v1.3.0)" sync_time=2023-04-27T16:20:29Z
2023-04-27T16:20:51Z INF waiting for source plugin to terminate module=cli
2023-04-27T16:20:51Z INF Got stop signal. Source plugin server shutting down address=/tmp/cq-doFdwiaueiZONRZD.sock module=cli signal=interrupt
2023-04-27T16:20:51Z INF Sending sync summary to analyticsv1.cloudquery.io:443 module=cli
```

### CloudQuery version

3.0.1

### Additional Context

_No response_

### Pull request (optional)

- [ ] I can submit a pull request

### PR DESCRIPTION
**Pull request by disq**
Fixes #10387

### CODE REVIEW

File: plugins/destination/file/client/spec.go [hunk 1]
```diff
@@ -32,8 +32,6 @@ func (s *Spec) SetDefaults() {
 		if !s.NoRotate {
 			s.Path += "." + PathVarUUID
 		}
-	} else {
-		s.Directory = path.Dir(s.Path)
```
Comment (disq) [inline plugins/destination/file/client/spec.go:36]: `s.Directory` is now not used in the processing logic, it's only used for deriving `Path` and printing a warning.

File: plugins/destination/file/client/write.go [hunk 2]
```diff
@@ -17,6 +18,11 @@ func (c *Client) WriteTableBatch(ctx context.Context, arrowSchema *arrow.Schema,
 	tableName := schema.TableName(arrowSchema)
 	timeNow := time.Now().UTC()
 	p := replacePathVariables(c.pluginSpec.Path, tableName, c.pluginSpec.Format, uuid.NewString(), timeNow)
+
+	if err := os.MkdirAll(path.Dir(p), 0755); err != nil {
```
Comment (disq) [inline plugins/destination/file/client/write.go:22]: This call was missing in the new 'path scheme', but it was creating directories with template variables in them in `client.New` (which was removed, above)

**Review by hermanschaaf — CHANGES_REQUESTED** [review]
Looks good, though can we modify (or add to) the tests so that the previous bug would have been caught?

**Review by candiduslynx — APPROVED** [review]
tests?

**Comment by disq** [comment]
> tests?

tests

File: plugins/destination/file/go.mod [hunk 3]
```diff
@@ -53,7 +53,12 @@ require (
 	gopkg.in/yaml.v2 v2.4.0 // indirect
 )
 
+require github.com/stretchr/testify v1.8.2
```
Comment (candiduslynx) [inline plugins/destination/file/go.mod:56]: move direct deps to 1 single place, indirect to another single place, go mod tidy?

**Review by candiduslynx — APPROVED** [review]

**Review by hermanschaaf — APPROVED** [review]
Nice, thanks for the fix!

File: plugins/destination/file/go.mod (hunk 3 above)
Comment (disq) [reply plugins/destination/file/go.mod]: Good idea, fixed

**Pull request merged**